src/main/java/org/tugraz/sysds/runtime/data/BasicTensor.java

```diff
@@ -564,31 +560,54 @@ public BasicTensor aggregateUnaryOperations(AggregateUnaryOperator op, BasicTens
 			dim1 = 2;
 		}
 		//prepare result matrix block
-		if(result==null || result._vt != _vt)
-			result = new BasicTensor(_vt, new int[]{dim0, dim1}, false);
-		else
-			result.reset(new int[]{dim0, dim1}, false);
+		BasicTensor res;
+		if(result==null || ((BasicTensor) result)._vt != _vt)
+			res = new BasicTensor(_vt, new int[]{dim0, dim1}, false);
+		else {
+			res = (BasicTensor) result;
+			res.reset(new int[]{dim0, dim1}, false);
+		}
 
 		if( LibTensorAgg.isSupportedUnaryAggregateOperator(op) )
 			if (op.indexFn instanceof ReduceAll)
-				LibTensorAgg.aggregateUnaryTensor(this, result, op);
+				LibTensorAgg.aggregateUnaryTensor(this, res, op);
 			else
 				throw new DMLRuntimeException("Only ReduceAll UnaryAggregationOperators are supported for tensor");
 		else
 			throw new DMLRuntimeException("Current UnaryAggregationOperator not supported for tensor");
-		return result;
+		return res;
 	}
 
-	public void incrementalAggregate(AggregateOperator aggOp, BasicTensor partialResult) {
+	public void incrementalAggregate(AggregateOperator aggOp, TensorBlock partialResult) {
 		if(!aggOp.correctionExists) {
 			if(aggOp.increOp.fn instanceof Plus) {
-				LibTensorAgg.aggregateBinaryTensor(partialResult, this, aggOp);
+				LibTensorAgg.aggregateBinaryTensor((BasicTensor) partialResult, this, aggOp);
 			}
 		}
 		else
 			throw new DMLRuntimeException("Correction not supported. correctionLocation: "+aggOp.correctionLocation);
 	}
 
+	@Override
+	public TensorBlock binaryOperations(BinaryOperator op, TensorBlock thatValue, TensorBlock result) {
+		if( !LibTensorBincell.isValidDimensionsBinary(this, thatValue) ) {
+			throw new RuntimeException("Block sizes are not matched for binary cell operations");
+		}
+		//prepare result matrix block
+		ValueType vt = resultValueType(_vt, ((BasicTensor) thatValue)._vt);
+		BasicTensor res;
```

Comment:
Could you maybe add a check that thatValue is indeed a BasicTensor? It would be great to have a single line similar to `BasicTensor that = checkType(thatValue, true);` which also avoids this potentially unchecked casting.